pmlc/dialect/tile/transforms/pad_ranges.cc

```diff
@@ -0,0 +1,148 @@
+      if (!lowerConst || lowerConst.getValue() != 0) {
+        op.emitRemark("Invalid lower bound");
+        return;
+      }
+      if (!upperConst) {
+        op.emitRemark("Invalid lower bound");
+        return;
+      }
+      ranges.push_back(upperConst.getValue() + 1);
+    }
+    // Get strides for output
+    auto outStrides = util::computeStrideArray(op.getResultType(), op.sink());
+    IVLOG(1, "outStrides = " << *outStrides);
+    // Decide wthat to round up
+    bool didChange = false;
+    SmallVector<int64_t, 6> newRanges = ranges;
+    for (unsigned i = 0; i < indexCount; i++) {
+      if (outStrides->strides[i] == 0) {
+        continue;
+      }
+      for (int64_t po2 = maxPowerOfTwo; po2 >= minPowerOfTwo; po2 /= 2) {
+        int64_t roundedRange = llvm::divideCeil(ranges[i], po2) * po2;
+        if (roundedRange == ranges[i]) {
+          break;
+        }
+        int64_t increase = roundedRange - ranges[i];
+        int64_t precIncrease = increase * 100 / ranges[i];
+        IVLOG(1, "Orig: " << ranges[i] << ", rounded: " << roundedRange
+                          << ", precIncrease = " << precIncrease);
+        if (precIncrease < maxIncreasePercent) {
+          newRanges[i] = roundedRange;
+          didChange = true;
+          break;
+        }
+      }
+    }
+    // Exit early if nothing is changing
+    if (!didChange) {
+      return;
+    }
+    IVLOG(1, "Making a brave new op");
+    // Start building a new contraction to 'shrink'
+    OpBuilder builder(op.getContext());
+    builder.setInsertionPointAfter(op.getOperation());
+    // Make shinking contraction
+    auto ident = createIdentity(builder, op.getLoc(), AtomicRMWKind::assign,
+                                op.getResultType().getElementType());
+    auto idMap = AffineMap::getMultiDimIdentityMap(outRank, op.getContext());
+    auto newOp = builder.create<ContractionOp>(
+        op.getLoc(), op.getResultType(), ident, ArrayRef<Value>{op},
+        AggregationKind::assign, CombinationKind::none, idMap,
+        ArrayRef<AffineMap>{idMap},
+        IntegerSet::getEmptySet(outRank, 0, op.getContext()),
+        /*noReduce=*/true, "shrink");
+    newOp.setLowerBounds(SmallVector<int64_t, 4>(outRank, 0));
+    SmallVector<int64_t, 4> newOutSize;
+    SmallVector<int64_t, 4> shinkUpperBounds;
+    for (unsigned i = 0; i < outRank; i++) {
+      newOutSize.push_back(newRanges[outIdx[i]]);
+      shinkUpperBounds.push_back(ranges[outIdx[i]] - 1);
+    }
+    newOp.setUpperBounds(shinkUpperBounds);
+    // Switch all uses to the new contracton
+    op.getResult().replaceAllUsesExcept(
+        newOp.getResult(), SmallPtrSet<Operation *, 1>{newOp.getOperation()});
+    // Resize + rerange the output of the original op
+    op.getResult().setType(
+        RankedTensorType::get(newOutSize, op.getResultType().getElementType()));
+    for (unsigned i = 0; i < indexCount; i++) {
+      newRanges[i] -= 1;
```

Comment:
Might use another variable with "bounds" in the name to make it clear you are converting from a range to bounds.

Reply:
Done.